src/promptflow/promptflow/executor/batch_engine.py

```diff
@@ -0,0 +1,98 @@
+        inputs_mapping: Dict[str, str],
+        output_dir: Path,
+        run_id: str = None,
+    ) -> BulkResult:
+        """Run flow in batch mode
+
+        :param input_dirs: The directories path of input files
+        :type input_dirs: Dict[str, str]
+        :param inputs_mapping: The mapping of input names to their corresponding values.
+        :type inputs_mapping: Dict[str, str]
+        :param output_dir: output dir
+        :type output_dir: The directory path of output files
+        :param run_id: The run id of this run
+        :type run_id: str
+        :return: The result of this batch run
+        :rtype: ~promptflow.executor._result.BulkResult
+        """
+        input_dicts = self._get_input_dicts(input_dirs, inputs_mapping)
+        batch_result = self.flow_executor.exec_bulk(input_dicts, run_id)
+        batch_result.outputs = self._persist_outputs(batch_result.outputs, output_dir)
+        return batch_result
+
+    def _get_input_dicts(self, input_dirs: Dict[str, str], inputs_mapping: Dict[str, str]):
+        """Resolve input data from input dirs and apply inputs mapping
+
+        TODO: After SDK/CLI can resolve input data from input dirs, this method can be removed.
+        """
+        input_dicts = self._resolve_data(input_dirs)
+        return self.flow_executor.validate_and_apply_inputs_mapping(input_dicts, inputs_mapping)
+
+    def _resolve_data(self, input_dirs: Dict[str, str]):
+        """Resolve input data from input dirs"""
+        result = {}
+        for input_key, input_dir in input_dirs.items():
+            input_dir = self._resolve_dir(input_dir)
+            file_data = load_data(input_dir)
+            for each_line in file_data:
+                self._resolve_image(input_dir, each_line)
+            result[input_key] = file_data
+        return result
+
+    def _resolve_dir(self, dir: Union[str, Path]) -> Path:
+        """Resolve input dir to absolute path"""
+        path = dir if isinstance(dir, Path) else Path(dir)
+        if not path.is_absolute():
+            path = self.flow_executor._working_dir / path
+        return path
+
+    def _resolve_image(self, input_dir: Path, one_line_data: dict):
+        """Resolve image path to absolute path in one line data"""
+        for key, value in one_line_data.items():
+            if isinstance(value, list):
+                for item in value:
+                    item = resolve_image_path(input_dir, item)
+                one_line_data[key] = value
+            elif isinstance(value, dict):
+                one_line_data[key] = resolve_image_path(input_dir, value)
+        return one_line_data
+
+    def _persist_outputs(self, outputs: List[Mapping[str, Any]], output_dir: Path):
+        """Persist outputs to output directory"""
+        output_dir = self._resolve_dir(output_dir)
+        # persist images to output directory
+        outputs = [persist_multimedia_data(output, output_dir) for output in outputs]
```

Comment:
You would need to persist output right after got result in `_timeout_process_wrapper` rather than doing it here. 
Doing it here would fill the memory with LineResult containing ImageBytes.

Also, all other image in LineResult that are not in LineResult.output should be replaced with string placeholder.

Eventually, we should not return BulkResult, we should return BulkSummary with statistics, status and error only, but that might took too much effort and delay the delivery of Image support. We need to assess the possibility.

Reply:
Updated.